### PR TITLE
Update VM state pre migration

### PR DESCRIPTION
**Pull request by stu-gott**
@rmohr

### CODE REVIEW

File: pkg/virt-controller/watch/pod.go [hunk 1]
```diff
@@ -88,21 +88,38 @@ func NewPodControllerWithListWatch(vmCache cache.Indexer, recorder record.EventR
...
-						// VM does not exist anymore, we don't have to retry
-						return true
-					}
-
-				}
-				logger.V(3).Info().Msg("Enqueuing VM initialization again.")
-				queue.AddRateLimited(key)
-				return true
-			}
+			if putVm(vm, restClient, queue, key) {return true}
```
Comment (rmohr) [inline pkg/virt-controller/watch/pod.go]: I would prefer

```golang
if err := putVM(); err != nil {
return true
}
```

File: pkg/virt-controller/watch/pod.go [hunk 2]
```diff
@@ -88,21 +88,38 @@ func NewPodControllerWithListWatch(vmCache cache.Indexer, recorder record.EventR
...
+		logger.Error().Reason(err).Msg("Setting the VM state failed.")
+		if e, ok := err.(*errors.StatusError); ok {
+			if e.Status().Reason == metav1.StatusReasonNotFound {
+				// VM does not exist anymore, we don't have to retry
+				return nil
+			}
+
+		}
+		logger.V(3).Info().Msg("Enqueuing VM again.")
+		queue.AddRateLimited(key)
```
Comment (rmohr) [inline pkg/virt-controller/watch/pod.go]: Could you move the queue logic out here back to the controller loop?

File: pkg/virt-controller/watch/pod.go [hunk 3]
```diff
@@ -87,22 +87,42 @@ func NewPodControllerWithListWatch(vmCache cache.Indexer, recorder record.EventR
...
-				queue.AddRateLimited(key)
+			logger := logging.DefaultLogger()
+			if putVm(vm, restClient, queue, key) {
+				return true
+			}
+			logger.Info().Msgf("VM successfully scheduled to %s.", vmCopy.Status.NodeName)
+		} else if vm.Status.Phase == corev1.Running {
+			vmCopy := corev1.VM{}
+			model.Copy(&vmCopy, vm)
+			vmCopy.Status.MigrationNodeName = vm.Status.NodeName
```
Comment (rmohr) [inline pkg/virt-controller/watch/pod.go]: That needs to be `vmCopy.Status.MigrationNodeName = pod.Status.NodeName`

File: pkg/virt-controller/watch/pod.go [hunk 4]
```diff
@@ -87,22 +87,42 @@ func NewPodControllerWithListWatch(vmCache cache.Indexer, recorder record.EventR
...
-				logger.V(3).Info().Msg("Enqueuing VM initialization again.")
-				queue.AddRateLimited(key)
+			logger := logging.DefaultLogger()
+			if putVm(vm, restClient, queue, key) {
+				return true
+			}
+			logger.Info().Msgf("VM successfully scheduled to %s.", vmCopy.Status.NodeName)
+		} else if vm.Status.Phase == corev1.Running {
+			vmCopy := corev1.VM{}
+			model.Copy(&vmCopy, vm)
```
Comment (rmohr) [inline pkg/virt-controller/watch/pod.go]: In the case of coping from one object to another object of the same type you can use the kubernetes copier `api.Scheme.Copy()`. It does the same thing but looks nicer ;)

File: pkg/virt-controller/watch/pod.go [hunk 5]
```diff
@@ -87,22 +87,42 @@ func NewPodControllerWithListWatch(vmCache cache.Indexer, recorder record.EventR
...
+			vmCopy.Status.MigrationNodeName = vm.Status.NodeName
+			logger := logging.DefaultLogger()
+			// TODO: the migration should be started here
```
Comment (rmohr) [inline pkg/virt-controller/watch/pod.go]: It should be started after the VM put.

File: pkg/virt-controller/watch/pod.go (hunk 2 above)
Comment (stu-gott) [inline pkg/virt-controller/watch/pod.go]: So we don't need to re-try on failure for the "Starting migration" case?

Comment (rmohr) [inline pkg/virt-controller/watch/pod.go]: We do. But I would prefer if you return the error and do the queue.AddRateLimited in the main controller loop.

File: pkg/virt-controller/watch/pod.go [hunk 6]
```diff
@@ -87,22 +87,42 @@ func NewPodControllerWithListWatch(vmCache cache.Indexer, recorder record.EventR
...
+		if e, ok := err.(*errors.StatusError); ok {
+			if e.Status().Reason == metav1.StatusReasonNotFound {
+				// VM does not exist anymore, we don't have to retry
+				return nil
+			}
+
+		}
+		logger.V(3).Info().Msg("Enqueuing VM again.")
+		queue.AddRateLimited(key)
+		return nil
```
Comment (rmohr) [inline pkg/virt-controller/watch/pod.go]: This shoud be `return err`

**Comment by stu-gott** [comment]
One test is currently failing. I've changed something with respect to the original pod scheduling behavior.

File: pkg/virt-controller/watch/pod.go [hunk 7]
```diff
@@ -87,22 +88,51 @@ func NewPodControllerWithListWatch(vmCache cache.Indexer, recorder record.EventR
...
+	if err := restClient.Put().Resource("vms").Body(&vm).Name(vm.ObjectMeta.Name).Namespace(kubeapi.NamespaceDefault).Do().Error(); err != nil {
+		logger.Error().Reason(err).Msg("Setting the VM state failed.")
+		if e, ok := err.(*errors.StatusError); ok {
+			if e.Status().Reason == metav1.StatusReasonNotFound {
+				// VM does not exist anymore, we don't have to retry
+				return nil
+			}
+
+		}
+		return fmt.Errorf("failed to set vm state: %v", err)
```
Comment (rmohr) [inline pkg/virt-controller/watch/pod.go]: Can't you just return the initial `err` here?

File: pkg/virt-controller/watch/pod.go [hunk 8]
```diff
@@ -87,22 +88,51 @@ func NewPodControllerWithListWatch(vmCache cache.Indexer, recorder record.EventR
...
+				queue.AddRateLimited(key)
+				return true
+			}
+			logger.Info().Msgf("VM successfully scheduled to %s.", vmCopy.Status.NodeName)
+		} else if vm.Status.Phase == corev1.Running {
+			logger := logging.DefaultLogger()
+			obj, err := kubeapi.Scheme.Copy(vm)
+			if err != nil {
+				logger.Error().Reason(err).Msg("could not copy vm object")
+				// FIXME: should the VM be re-enqueued? failure to copy is a pretty big deal
```
Comment (rmohr) [inline pkg/virt-controller/watch/pod.go]: Not sure myself here. Probably rate limitting a few times and then dismiss the key and forget about it's enqueue history ... (queue.Forget())

File: pkg/virt-controller/watch/pod.go [hunk 9]
```diff
@@ -87,22 +88,51 @@ func NewPodControllerWithListWatch(vmCache cache.Indexer, recorder record.EventR
...
 
-				}
-				logger.V(3).Info().Msg("Enqueuing VM initialization again.")
+			if err := putVm(vmCopy, restClient, queue); err != nil {
+				logger.V(3).Info().Msg("Enqueuing VM again.")
 				queue.AddRateLimited(key)
 				return true
 			}
+			// TODO: the migration should be started here
 			logger.Info().Msgf("VM successfully scheduled to %s.", vmCopy.Status.NodeName)
```
Comment (rmohr) [inline pkg/virt-controller/watch/pod.go]: This message is wrong ..

**Comment by stu-gott** [comment]
Failing test was caused by erroneously passing a pointer by reference

**Pull request merged**